parla/parray/coherence.py

```diff
@@ -0,0 +1,241 @@
+from __future__ import annotations # To support self references in type checking, must be first line
+
+from typing import List
+
+class MemoryOperation:
+    """
+    A memory operation representation.
+    """
+    ERROR = -1  # there is an error
+    NOOP = 0    # no operation
+    LOAD = 1    # load data from src to dst
+    EVICT = 2   # clear the data in src
+
+    def __init__(self, inst: int = NOOP, dst: int = -1, src: int = -1):
+        self.inst = inst
+        self.dst = dst
+        self.src = src
+
+    @staticmethod
+    def noop() -> MemoryOperation:
+        return MemoryOperation()
+
+    @staticmethod
+    def error() -> MemoryOperation:
+        return MemoryOperation(MemoryOperation.ERROR)
+
+    @staticmethod
+    def load(dst, src) -> MemoryOperation:
+        return MemoryOperation(MemoryOperation.LOAD, dst, src)
+
+    @staticmethod
+    def evict(src) -> MemoryOperation:
+        return MemoryOperation(MemoryOperation.EVICT, src=src)
+
+
+class Coherence:
+    """
+    A memory coherence protocol between devices.
+
+    Implements MSI protocol.
+    """
+    INVALID = 0
+    SHARED = 1
+    MODIFIED = 2
+
+    def __init__(self, init_owner: int):
+        """
+        Args:
+            init_owner: the owner of the first copy in the system
+        """
+        self._coherence_states = {init_owner: self.SHARED}  # states of each device
+        self._owner = init_owner  # owner id when MODIFIED / smallest valid device id when SHARED

```

Comment:
How is owner used?

Reply:
owner the where the data should be loaded from. If there are multiple devices share a copy (SHARED state), the smallest one (so likely to be CPU), will becomes the owner. And later other INVALID device will read from it.

Reply:
Sounds good. might want to not make it CPU bound, since device->device is usually faster, maybe `max` instead. Or possibly load from different locations assuming in SHARED state, if two auto-moves to new devices are called at once, to prevent bottlenecking. 

But these are later tiny optimizations that likely don't matter much with all our other otherheads.

Reply:
Oh I thought PCIe is faster, looks like I am wrong.

Reply:
It depends if the system has P2P or if the cpu buffer is already pinned. Most HPC workstations do have P2P now though, so it generally should be faster.